mumuki/mumuki-laboratory
# Feature puzzle

# :dart: Goal

:jigsaw: This PR add the necessary support for implementing puzzles in the platform. 

# :memo: Details

> ~This PR add the necessary support for implementing puzzles in the platform~

Well, nope. This PR does not actually add support for puzzles :exploding_head: , but improves the support for `custom` editors instead. 

Custom editors were born at the same time kids layout and blocks support were introduced. That made the original implementation of this three features, although keeping separate was a design goal, and during the following development months that goal was achieved, at some extend - see #1425.

This PR improves this support because: 

* It fully documents - both inline and in README - the hooks required to implement custom editors
* Removes a few kids-specific references
* Documents some types using JSDoc
* Introduces the `mumuki.submission.syncContent` API, which decouples content syncing from codemirror - which is a design flaw from early 2016. Although it does no fix the real issue - all editors are calling codemirror-specific code under the hood even when they don't use - it prevents perpetuating it in custom editors and opens the door to fixing it. 
* `mumuki.submission.getContent` has been exposed and rewritten, so that `CustomEditor.addSource` takes precedence over it. 
* `mumuki.submission.processSolution` API was introduced, that allows layout-agnostic programatic submission of solutions. Before this change, you need to manually call kids-specific functions in order to update UI, and there was no submit-render-and-all-the-other-stuff API

# :rofl: Fun facts

This PR is called feature-puzzle, but there is no single reference to puzzles in the code :stuck_out_tongue_closed_eyes: nor in is comments. 

# :back: Backward compatibilty

This PR is theoretically not backwards compatible, since it clears `CustomEditor` sources on init, but only scratch :cat: runner is using it - https://github.com/mumuki/mumuki-scratch-runner/blob/master/lib/render/editor/editor.js#L38 -, and it looks to be called in a `document.onReady` block of a non-turbolinks code, so it should be ok and everything should work. 

The same applies to the `getContent` changes. 

# :shrug: Known issues

By just following the documented steps the reset button in kids will break. I will dig on that in the future

# :test_tube:  Test instructions

1. Start puzzle runner

```bash
bundle exec rackup
```

2. Seed database with a new puzzle exercise: 

```ruby
language = Mumuki::Domain::Seed.languages_syncer.locate_and_import!  Language, 'http://localhost:9292'
e = Exercise.find(2072).dup
e.number = e.guide.exercise.count 
e.editor = :custom
e.language = language
e.description = "Llegó el momento de conocer al pintor Xul Solar.\n\n > ¡Ahora te toca a vos! Armá el rompecabezas"
e.initial_state = '<img src="https://flbulgarelli.github.io/headbreaker/static/xul.jpg">'
e.expectations = []
e.test = '
// this first call will not be necessary in the future
Muzzle.expect([
  [1,1],[2,1],[3,1],[4,1],
  [1,2],[2,2],[3,2],[4,2],
  [1,3],[2,3],[3,3],[4,3],
  [1,4],[2,4],[3,4],[4,4]
]);
Muzzle.basic(4, 4, "https://flbulgarelli.github.io/headbreaker/static/xul.jpg");
'
e.layout = :input_bottom
e.save!
e.id
```
3.  Point browser to `http://localhost:3000/central/exercises/` + last printed it
4. Repeat the process using layouts `e.layout = :input_kindergarten` and `e.layout = :input_primary`

# :eyes: See also

* :factory_worker:   https://github.com/mumuki/mumuki-puzzle-runner: the on-going puzzle runner that uses all this new features

# :camera:  Screenshots

## Input Bottom

![image](https://user-images.githubusercontent.com/677436/87177732-81733600-c2b2-11ea-95fa-dd4b0eaf44e1.png)

![image](https://user-images.githubusercontent.com/677436/87256161-ba540c00-c466-11ea-99be-052425aca8b2.png)

## Input Primary

![image](https://user-images.githubusercontent.com/677436/87177629-5daff000-c2b2-11ea-8a29-4a2d2e7145f7.png)


## Input Kindergarten 

![image](https://user-images.githubusercontent.com/677436/87177526-2b9e8e00-c2b2-11ea-8e23-24d43320b81e.png)

![image](https://user-images.githubusercontent.com/677436/87256576-4e73a280-c46a-11ea-824c-bb795e55a20d.png)

![image](https://user-images.githubusercontent.com/677436/87256586-63e8cc80-c46a-11ea-958f-bd55f8804556.png)
